pkg/restore/change_pvc_node_selector.go

```diff
@@ -0,0 +1,156 @@
+// if a mapping is found in the plugin's config map.
+type ChangePVCNodeSelectorAction struct {
+	logger          logrus.FieldLogger
+	configMapClient corev1client.ConfigMapInterface
+	nodeClient      corev1client.NodeInterface
+}
+
+// NewChangePVCNodeSelectorAction is the constructor for ChangePVCNodeSelectorAction.
+func NewChangePVCNodeSelectorAction(
+	logger logrus.FieldLogger,
+	configMapClient corev1client.ConfigMapInterface,
+	nodeClient corev1client.NodeInterface,
+) *ChangePVCNodeSelectorAction {
+	return &ChangePVCNodeSelectorAction{
+		logger:          logger,
+		configMapClient: configMapClient,
+		nodeClient:      nodeClient,
+	}
+}
+
+// AppliesTo returns the resources that ChangePVCNodeSelectorAction should be run for
+func (p *ChangePVCNodeSelectorAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"persistentvolumeclaims"},
+	}, nil
+}
+
+// Execute updates the pvc's selected-node annotation:
+//    a) if node mapping found in the config map for the plugin
+//	  b) if node mentioned in annotation doesn't exist
+func (p *ChangePVCNodeSelectorAction) Execute(input *velero.RestoreItemActionExecuteInput) (*velero.RestoreItemActionExecuteOutput, error) {
+	p.logger.Info("Executing ChangePVCNodeSelectorAction")
+	defer p.logger.Info("Done executing ChangePVCNodeSelectorAction")
+
+	typeAcc, err := meta.TypeAccessor(input.Item)
+	if err != nil {
+		return &velero.RestoreItemActionExecuteOutput{}, err
+	}
+
+	metadata, err := meta.Accessor(input.Item)
+	if err != nil {
+		return &velero.RestoreItemActionExecuteOutput{}, err
+	}
+
+	annotations := metadata.GetAnnotations()
+	if annotations == nil {
+		return velero.NewRestoreItemActionExecuteOutput(input.Item), nil
+	}
+
+	log := p.logger.WithFields(map[string]interface{}{
+		"kind":      typeAcc.GetKind(),
+		"namespace": metadata.GetNamespace(),
+		"name":      metadata.GetName(),
+	})
+
+	// let's check if PVC has annotation of the selected node
+	node, ok := annotations["volume.kubernetes.io/selected-node"]
+	if !ok {
+		log.Debug("PVC doesn't have node selector")
+		return velero.NewRestoreItemActionExecuteOutput(input.Item), nil
+	}
+
+	// fetch node mapping from configMap
+	newNode, err := getNewNodeFromConfigMap(p.configMapClient, node)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(newNode) != 0 {
+		// set node selector
+		// We assume that node exist for node-mapping
+		annotations["volume.kubernetes.io/selected-node"] = newNode
+		metadata.SetAnnotations(annotations)
+		log.Infof("Updating selected-node to %s from %s", newNode, node)
+		return velero.NewRestoreItemActionExecuteOutput(input.Item), nil
+	}
+
+	// configMap doesn't have node-mapping
+	// Let's check if node exists or not
+	exists, err := isNodeExist(p.nodeClient, node)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error checking node %s existence", node)
+	}
+
+	if !exists {
+		log.Infof("Resetting selected-node")
```

Comment:
Maybe "Clearing selected-node because node named %s does not exist" would be a little clearer